flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java

```diff
@@ -0,0 +1,241 @@
+		super(path);
+		this.fieldTypes = fieldTypes;
+		this.fieldNames = fieldNames;
+		// read whole parquet file as one file split
+		this.unsplittable = true;
+	}
+
+	@Override
+	public Tuple2<Long, Long> getCurrentState() {
+		return new Tuple2<>(this.lastSyncedBlock, this.recordsReadSinceLastSync);
+	}
+
+	@Override
+	public void open(FileInputSplit split) throws IOException {
+		org.apache.hadoop.conf.Configuration configuration = new org.apache.hadoop.conf.Configuration();
+		InputFile inputFile =
+			HadoopInputFile.fromPath(new org.apache.hadoop.fs.Path(split.getPath().toUri()), configuration);
+		ParquetReadOptions options = ParquetReadOptions.builder().build();
+		ParquetFileReader fileReader = new ParquetFileReader(inputFile, options);
+		MessageType schema = fileReader.getFileMetaData().getSchema();
+		this.skipThisSplit = false;
+		MessageType readSchema = getReadSchema(schema);
+		this.parquetRecordReader = new ParquetRecordReader<>(new RowReadSupport(), readSchema, FilterCompat.NOOP);
+		this.parquetRecordReader.initialize(fileReader, configuration);
+		if (this.recordConsumed == null) {
+			this.recordConsumed = getRuntimeContext().getMetricGroup().counter("parquet-records-consumed");
+		}
+
+		if (!skipThisSplit) {
+			LOG.debug(String.format("Open ParquetInputFormat with FileInputSplit [%s]", split.getPath().toString()));
+		} else {
+			LOG.warn(String.format(
+				"Escaped the file split [%s] due to mismatch of file schema to expected result schema",
+				split.getPath().toString()));
+		}
+	}
+
+	@Override
+	public void reopen(FileInputSplit split, Tuple2<Long, Long> state) throws IOException {
+		Preconditions.checkNotNull(split, "reopen() cannot be called on a null split.");
+		Preconditions.checkNotNull(state, "reopen() cannot be called with a null initial state.");
+
+		try {
+			this.open(split);
+		} finally {
+			if (getCurrentState().f0 != -1) {
+				lastSyncedBlock = state.f0;
+				recordsReadSinceLastSync = state.f1;
+			}
+		}
+
+		if (lastSyncedBlock != -1) {
+			// open and read util the record we were before
+			// the checkpoint and discard the values
+			parquetRecordReader.seek(lastSyncedBlock);
+			for (int i = 0; i < recordsReadSinceLastSync; i++) {
+				// skip the record already processed
+				parquetRecordReader.nextRecord();
+			}
+		}
+	}
+
+	protected String[] getFieldNames() {
+		return fieldNames;
+	}
+
+	protected TypeInformation[] getFieldTypes() {
+		return fieldTypes;
+	}
+
+	@Override
+	public void close() throws IOException {
+		if (parquetRecordReader != null) {
+			parquetRecordReader.close();
+		}
+	}
+
+	@Override
+	public boolean reachedEnd() throws IOException {
+		if (skipThisSplit) {
+			return true;
+		}
+		return parquetRecordReader.reachEnd();
+	}
+
+	@Override
+	public E nextRecord(E e) throws IOException {
+		if (reachedEnd()) {
+			return null;
+		}
+
+		// Means start to create a new block
+		if (parquetRecordReader.getCurrentBlock() != lastSyncedBlock) {
```

Comment:
I think it would be good if we could move all logic to keep track of the read position and reset the read position into the `ParquetRecordReader`. That means, we do not need to keep track of read records in this class, but just fetch them from the record reader when we need to take a checkpoint.